lib/std/os/windows.zig

```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
+) ![:0]u16 {
+    const alignment = @alignOf(OBJECT_NAME_INFORMATION);
+    const begin_addr = @ptrToInt(out_buffer.ptr);
+    const aligned_begin_addr = std.mem.alignForward(begin_addr, alignment);
+    const alignment_step = aligned_begin_addr - begin_addr;
+    const aligned_raw_buffer = @alignCast(alignment, std.mem.sliceAsBytes(out_buffer)[alignment_step..out_buffer.len*@sizeOf(u16)]);
+    if(aligned_raw_buffer.len < @sizeOf(OBJECT_NAME_INFORMATION)) return error.BufferTooSmall; //this would result in .INFO_LENGTH_MISMATCH
+    
+    var info = @ptrCast(*OBJECT_NAME_INFORMATION, aligned_raw_buffer);
+    info.* = undefined; //all of the fields we read will be overwritten, and the result string is placed in the same buffer after the struct
+    const out_buffer_length = std.math.cast(ULONG, aligned_raw_buffer.len) catch |e| std.math.maxInt(u32); //we'll just use as much of it as we can
+    const rc = ntdll.NtQueryObject(Handle, .ObjectNameInformation, out_buffer.ptr, out_buffer_length, ReturnLength); //buffer size is specified in u16...
+    switch(rc){
+        .SUCCESS => if(info.Name.Buffer) |buffer| return buffer[0..@divExact(info.Name.Length, 2):0] //... but result length is specified in u8
+            else return error.Unnamed,
+        .ACCESS_DENIED => return error.AccessDenied,
+        .INVALID_HANDLE => return error.InvalidHandle,
+        .BUFFER_OVERFLOW, .BUFFER_TOO_SMALL => return error.NameTooLong,
+        .INFO_LENGTH_MISMATCH => unreachable, //aligned_raw_buffer.len < @sizeOf(OBJECT_NAME_INFORMATION) already checked previously
+        else => |e| return unexpectedStatus(e),
+    }
+}
+test "QueryObjectName" {
+    if (comptime builtin.os.tag != .windows)
+        return;
+    
+    const file: std.fs.File = try std.fs.openSelfExe(.{}); //any file will do; canonicalization works on NTFS junctions and symlinks, hardlinks remain separate paths.
+    defer file.close();
+    var out_buffer align(16) = std.mem.zeroes([1<<10]u16); //make this big enough for the test runner exe path
+    var required_length: ULONG = undefined;
+
+    _ = try QueryObjectName(file.handle, out_buffer[0..], &required_length);
+    required_length = @divExact(required_length, 2);
+    _ = try QueryObjectName(file.handle, out_buffer[0..required_length], null);
+    std.testing.expect(null == QueryObjectName(file.handle, out_buffer[0..required_length-1], null)
+        catch |e| switch(e) {error.NameTooLong => null, else => unreachable});
+    _ = try QueryObjectName(file.handle, out_buffer[0..required_length+1], null);
+}
+
+pub const GetFinalPathNameByHandleError = error {
+        BadPathName,
+        FileNotFound,
+        NameTooLong,
+        Unexpected,
+    }
+    || if((comptime builtin.os.tag != .windows) or (targetVersionIsAtLeast(WindowsVersion.win10_rs4) == true))
+        error {}
+    else
+        error {
+            AccessDenied,
+            BufferTooSmall,
+            InvalidHandle,
+            Unnamed,
```

Comment:
Unnamed ?

Reply:
That possibility is mentioned [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-obquerynamestring) (`if the given object is unnamed`), which I found used in the implementation of ReactOS's `NtQueryObject`. I don't know when it would be triggered though, might be for other "objects" than files? I'll change that to unreachable.

Reply:
Hmm, could we not coerce `Unnamed` as simply `Unexpected`? Or put it another way, what value will conveying that the object is unnamed give to the user?